### PR TITLE
[8/x][programmable transactions] Added MakeMoveVec

### PR DESCRIPTION
**Pull request by tnowacki**
## Description 

- MakeMoveVec is the more general alternative to CallArg::ObjVec
- Built on #8633 

## Test Plan 

No tests yet

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-types/src/messages.rs [hunk 1]
```diff
@@ -449,10 +453,13 @@ impl Command {
                 ));
                 fp_ensure!(!is_blocked, UserInputError::BlockedMoveFunction);
             }
-            Command::TransferObjects(_, _)
-            | Command::SplitCoin(_, _)
-            | Command::MergeCoins(_, _)
-            | Command::Publish(_) => (),
+            Command::TransferObjects(args, _) | Command::MergeCoins(_, args) => {
```
Comment (tnowacki) [inline crates/sui-types/src/messages.rs:456]: Unrelated fix

File: crates/sui-types/src/messages.rs [hunk 2]
```diff
@@ -480,13 +487,19 @@ impl ProgrammableTransaction {
             .iter()
             .flat_map(|arg| arg.input_objects())
             .collect::<Vec<_>>();
+        // all objects, not just mutable, must be unique
         let mut used = HashSet::new();
         if !input_arg_objects.iter().all(|o| used.insert(o.object_id())) {
             return Err(UserInputError::DuplicateObjectRefInput);
         }
+        // do not duplicate packages referred to in commands
+        let command_input_objects: BTreeSet<InputObjectKind> = commands
```
Comment (tnowacki) [inline crates/sui-types/src/messages.rs:496]: Unrelated fix

File: crates/sui-adapter/src/programmable_transactions/context.rs [hunk 3]
```diff
@@ -578,6 +592,87 @@ where
...
+
+    pub fn join(&self, t1: ValueType, t2: ValueType) -> Result<Option<ValueType>, ExecutionError> {
+        Ok(match (t1, t2) {
+            // everything joins with itValueType
+            (t1, t2) if t1 == t2 => Some(t1),
+            // AnyPrimitive joins with primitive types types. It is otherwise unsafe to join with
+            // AnyPrimitive as it is just a series of BCS bytes
+            // TODO dev-inspect, allow AnyPrimitive to join with any type?
+            (ValueType::AnyPrimitive, ValueType::Loaded { ty, abilities })
+            | (ValueType::Loaded { ty, abilities }, ValueType::AnyPrimitive) => {
```
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/context.rs:649]: This case scares me a bit but maybe feels okay? Basically its a question of if we want to let `CallArg::Pure` appear in `MakeMoveVec`

File: crates/sui-adapter/src/programmable_transactions/types.rs [hunk 4]
```diff
@@ -271,23 +310,23 @@ fn try_from_value_prim<'a, T: Deserialize<'a>>(
...
             if ty == &expected_ty {
                 panic!("type mismatch")
             }
-            let res = bcs::from_bytes(bytes);
-            assert_invariant!(
-                res.is_ok(),
-                "Move values should be able to deserialize into their type"
-            );
-            Ok(res.unwrap())
+            let Ok(res) = bcs::from_bytes(bytes) else {
```
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/types.rs:326]: MakeMoveVec generates loaded, so this is no longer guaranteed to succeed

File: crates/sui-adapter/src/programmable_transactions/execution.rs [hunk 5]
```diff
@@ -76,6 +91,55 @@ fn execute_command<E: fmt::Debug, S: StorageView<E>>(
...
     let results = match command {
+        Command::MakeMoveVec(args) if args.is_empty() => {
+            vec![Value::EmptyVec]
+        }
+        Command::MakeMoveVec(args) => {
+            let mut res = vec![];
+            leb128::write::unsigned(&mut res, args.len() as u64).unwrap();
+            let mut arg_iter = args.into_iter().enumerate();
+            let (idx, first_arg) = arg_iter.next().unwrap();
+            let first_value: Value = context.take_arg(CommandKind::MakeMoveVec, idx, first_arg)?;
```
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:102]: A bit strange here, as we cannot determine ahead of time to clone or take

File: crates/sui-adapter/src/programmable_transactions/execution.rs [hunk 6]
```diff
@@ -76,6 +91,55 @@ fn execute_command<E: fmt::Debug, S: StorageView<E>>(
...
+                    type_ = joined
+                } else {
+                    panic!("type mismatch")
+                }
+                used_in_non_entry_move_call =
+                    used_in_non_entry_move_call || value.was_used_in_non_entry_move_call();
+                res.extend(value.to_bcs_bytes());
+            }
+            let inner_ty = match type_ {
+                ValueType::AnyPrimitive | ValueType::AnyVec => panic!(
```
Comment (tnowacki) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:118]: @amnn I'll probably need your opinion here as you have been following the runtime closely. 

This case is also a bit strange, and with the clone vs take above, I wonder if `MakeMoveVec` should just take a type tag? 
Pro:
- No Value::EmptyVec
- No Weird edge case here
- we can check for primitives first which would simplify the join case for `AnyPrimitive`
Con: 
- Makes it not possible to migrate `CallArg::ObjVec` naively, as the type tag would now need to be specified.

Comment (tnowacki) [reply crates/sui-adapter/src/programmable_transactions/execution.rs]: I went ahead and added the type argument, which solves a lot of issues. I made it optional for non-empty vector of objects, just to make the migration a tad easier... but hopefully that isn't confusing.

Comment (amnn) [reply crates/sui-adapter/src/programmable_transactions/execution.rs]: +1 to including the type tag at least for the empty vector case -- the behaviour of an empty vector (w.r.t. droppability) changes based on its type, so it makes sense to me that it would need to be specified.  The other option is to have `ObjVec` and `PureVec` (where "pure" implies "copy + drop")?

Comment (tnowacki) [reply crates/sui-adapter/src/programmable_transactions/execution.rs]: I tried a few different approaches and like where I landed (I think)
- If it is empty, you must specify a type tag
- If the elements are objects, you _don't_ have to specify a type tag
- If the elements are not objects, clane/take is decided by the abilities of the type tag

File: crates/sui-adapter/src/programmable_transactions/execution.rs [hunk 7]
```diff
@@ -93,6 +93,78 @@ fn execute_command<E: fmt::Debug, S: StorageView<E>>(
...
+            leb128::write::unsigned(&mut res, args.len() as u64).unwrap();
+            let mut arg_iter = args.into_iter().enumerate();
+            let (mut used_in_non_entry_move_call, tag) = match tag_opt {
+                Some(tag) => (false, tag),
+                // If no tag specified, it _must_ be an object
+                None => {
+                    // empty args covered above
+                    let (idx, arg) = arg_iter.next().unwrap();
+                    let obj: ObjectValue = context.take_arg(CommandKind::MakeMoveVec, idx, arg)?;
+                    res.extend(obj.to_bcs_bytes());
```
Comment (amnn) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:130]: nit: Is there a way to push bcs bytes for a value onto an existing byte buffer?

File: crates/sui-adapter/src/programmable_transactions/execution.rs [hunk 8]
```diff
@@ -93,6 +93,78 @@ fn execute_command<E: fmt::Debug, S: StorageView<E>>(
...
+            let elem_abilities = context
+                .session
+                .get_type_abilities(&elem_ty)
+                .map_err(|e| context.convert_vm_error(e))?;
+            for (idx, arg) in arg_iter {
+                let value: Value = if elem_abilities.has_copy() {
+                    context.clone_arg(idx, arg)?
+                } else {
+                    context.take_arg(CommandKind::MakeMoveVec, idx, arg)?
+                };
```
Comment (amnn) [inline crates/sui-adapter/src/programmable_transactions/execution.rs:148]: It's this kind of conditional that I was trying to avoid with one of the suggestions I made on an earlier PR: I think there should only be `context.take_arg`, and that should check whether the arg at the given index has copy, and behave like `clone` if it does, and take if not.

**Review by amnn — APPROVED** [review]
Change LGTM!

File: crates/sui-adapter/src/programmable_transactions/execution.rs (hunk 8 above)
Comment (tnowacki) [reply crates/sui-adapter/src/programmable_transactions/execution.rs:148]: Will fix in the next PR

File: crates/sui-adapter/src/programmable_transactions/execution.rs (hunk 7 above)
Comment (tnowacki) [reply crates/sui-adapter/src/programmable_transactions/execution.rs:130]: Will fix in the next PR
